torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,354 @@
+concatenate all metadata values with a feature index and use this as a unique
+index in a KV store. More complicated implementations may choose to partition
+features in interesting manners based on the provided metadata.
+
+Major TODOs for future implementation:
+* Async `put` and `get` functionality
+"""
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Optional, Union
+
+import numpy as np
+import torch
+
+from torch_geometric.typing import FeatureTensorType
+from torch_geometric.utils.mixin import CastMixin
+
+_field_status = Enum("FieldStatus", "UNSET")
+
+IndexType = Union[FeatureTensorType, slice]
+
+
+@dataclass
+class TensorAttr(CastMixin):
+    r"""Defines the attributes of a :obj:`FeatureStore` tensor; in particular,
+    all the parameters necessary to uniquely identify a tensor from the feature
+    store.
+
+    Note that the order of the attributes is important; this is the order in
+    which attributes must be provided for indexing calls. Feature store
+    implementor classes can define a different ordering by overriding
+    TensorAttr.__init__.
+    """
+
+    # The type of the nodes that the tensor corresponds to (may be used for
+    # hetereogeneous graphs)
+    group_name: Optional[str] = _field_status.UNSET
+
+    # The name of the feature tensor (may be used if there are multiple
+    # different feature tensors for the same node index)
+    attr_name: Optional[str] = _field_status.UNSET
+
+    # The node indices the rows of the tensor correspond to
+    index: Optional[IndexType] = _field_status.UNSET
+
+    # Convenience methods #####################################################
+
+    def is_set(self, attr):
+        r"""Whether an attribute is set in :obj:`TensorAttr`."""
+        assert attr in self.__dataclass_fields__
+        return getattr(self, attr) != _field_status.UNSET
+
+    def is_fully_specified(self):
+        r"""Whether the :obj:`TensorAttr` has no unset fields."""
+        return all([
+            getattr(self, field) != _field_status.UNSET
+            for field in self.__dataclass_fields__
+        ])
+
+    def update(self, attr: 'TensorAttr'):
+        r"""Updates an :obj:`TensorAttr` with set attributes from another
+        :obj:`TensorAttr`."""
+        for field in self.__dataclass_fields__:
+            val = getattr(attr, field)
+            if val != _field_status.UNSET:
+                setattr(self, field, val)
+
+
+class AttrView(CastMixin):
+    r"""Defines a view of a :obj:`FeatureStore` that is obtained from a
+    specification of attributes on the feature store. The view stores a
+    reference to the backing feature store as well as a :obj:`TensorAttr`
+    object that represents the view's state.
+
+    Users can create views either using the :obj:`AttrView` constructor,
+    :obj:`FeatureStore.view`, or by incompletely indexing a feature store.
```

Comment:
This is super cool. Maybe consider adding some `code-block:: python` example directly in this doc-string.